File: NodeLink/61. 旋转链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
 var rotateRight = function(head, k) {
    // 排除空链表
    if(head == null) return head
    // 设置p指针 将p指针移到最后一个节点
    let p = head;
    // 记录总节点数
    let n = 1
    while(p.next !== null){
       p  = p.next 
       n++;
    } 
    // 将链表变成循环链表
    p.next = head
    // 如果k过大 对n 取余,减少循环次数
    k %= n
    // 以为是循环左右,将循环字数变为循环于n的差,使用循环右移解决
    k = n - k;
    // 循环k次
    while(k--){
        p = p.next
    }
    // 因为从最后节点开始循环,所以p节点的下一个节点为,循环的头部
    head = p.next;
    p.next = null
    return head;

};

File: Python/61. 旋转链表.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-6 10:41
# @Author : 刘永奇
# @File : 61. 旋转链表.py
# @Software : PyCharm
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next: return head
        n = 1
        p = head
        while p.next:
            p = p.next
            n += 1
        p.next = head
        k %= n
        l = n - k
        while l:
            p = p.next
            l -= 1
        res = p.next
        p.next = None
        return res


